memory-viz/src/user_functions.ts

```diff
@@ -45,54 +56,75 @@ function draw(
         objs = objects;
     }
 
-    let m: MemoryModel;
+    const isArrayOfArrays = Array.isArray(objs) && Array.isArray(objs[0]);
 
-    if (automation) {
-        m = drawAutomated(objs, configuration.width, configuration);
-    } else {
-        // Dynamically determining the width of the canvas, in case one has not been provided.
-        if (!configuration.hasOwnProperty("width")) {
-            let rightmost_obj;
-            let rightmost_edge = 0;
+    const processSnapshot = (snapshotObjects: DrawnEntity[]) => {
+        getCanvasDimensions(configuration, snapshotObjects);
+        const model = new MemoryModel({
+            width: configuration.width,
+            height: configuration.height,
+            roughjs_config: configuration.roughjs_config,
+        });
+        model.drawAll(snapshotObjects);
+        return model;
+    };
 
-            for (const obj of objs) {
-                const width = getSize(obj).width;
-                const curr_edge = obj.x + width;
-                if (curr_edge > rightmost_edge) {
-                    rightmost_edge = curr_edge;
-                    rightmost_obj = obj;
-                }
+    if (isArrayOfArrays) {
+        const snapshots = objs as DrawnEntity[][];
+        return automation
+            ? snapshots.map((snapshotObjects) =>
+                  drawAutomated(
+                      snapshotObjects,
+                      configuration.width,
+                      configuration
+                  )
+              )
+            : snapshots.map(processSnapshot);
+    }
+
+    const snapshotObjects = objs as DrawnEntity[];
+    return automation
+        ? drawAutomated(snapshotObjects, configuration.width, configuration)
+        : processSnapshot(snapshotObjects);
+}
+
+function getCanvasDimensions(
+    configuration: Partial<DisplaySettings>,
+    snapshotObjects: DrawnEntity[]
+): void {
```

Comment:
@david-yz-liu I extracted this code for getting the canvas dimension in the manual mode. I was wondering if it would be better to move it into MemoryModel?

Reply:
Hmmm yes, I think it's better to move into `MemoryModel`. Good suggestion